### DISCUSSION TITLE
ScriptReference/MenuItem

### DISCUSSION THREAD

**Discussion by giscus** [Developer-Notes-Extension/Unity-Notes#34] [Notes]
# ScriptReference/MenuItem

https://docs.unity3d.com/ScriptReference/MenuItem.html

**Comment by vertxxyz** [comment]
If you're using simple shortcuts with shift key alone (eg. <kbd>Shift+x</kbd>) that keyboard event will be consumed regardless of whether you have a validation method or not.
So when you're inevitably focused in a Text Field or Area, you can no longer type that capital letter.
My workaround involves this in the validation function for your shortcut:

```csharp
if (EditorGUIUtility.editingTextField)
{
	EditorWindow focused = focusedWindow;
	if (focused == null)
		return false;
	Event keyboardEvent = Event.KeyboardEvent("X");
	keyboardEvent.character = 'X'; //The variable that TextFields seem to actually use.
	keyboardEvent.modifiers = EventModifiers.Shift;
	keyboardEvent.shift = true;
	focused.SendEvent(keyboardEvent);
	return false;
}
return true;
```

**Comment by vertxxyz** [comment]
You can use [`EditorApplication.ExecuteMenuItem`](https://docs.unity3d.com/ScriptReference/EditorApplication.ExecuteMenuItem.html) to manually invoke a MenuItem by path.
You can use [`Menu.SetChecked`](https://docs.unity3d.com/ScriptReference/Menu.SetChecked.html) to set the check-box state of a menu item.

**Comment by cmcpasserby** [comment]
When you use this via the context menu with multiple objects selected it will execute multiple times.  
You can use the `MenuCommand.context` to see what object it is acting on and compare with your selection if you want it to only execute once.  
MenuCommand.context will be null if the menu item is executed from the top menu or keyboard shortcut.  
  

```csharp
[MenuItem("GameObject/How Many Objects Are Selected")]  
private static void HowManyObjectsAreSelected(MenuCommand menuCommand) {  
    if (Selection.gameObjects.Length > 1 && menuCommand.context != Selection.gameObjects[0]) {  
        return;  
    }  
    Debug.Log($"You have {objects.Length} objects selected");  
}
```

**Comment by feranti** [comment]
[Source](https://hugecalf-studios.github.io/unity-lessons/lessons/editor/menuitem/)

It is very simple to disable the use of a certain menu item, yet the Unity docs are not clear. To disable a menu item, we must make use of what Unity called a Validation Function. A Validation Function is a static bool method which gets executed before Unity displays your custom menu to you. To make a disabled menu item, simply return false from this method. Conversely, returning true (or not having a Validation Function) will cause the menu item to be enabled.

The confusion comes in by assigning the Validation Function. To do so, you must use the exact same MenuItem attribute as before, with the exact same path given as the first parameter. This tells Unity which menu item to apply this Validation Function to. You must then give true as the second parameter, which tells Unity that this is a Validation Function.

![Custom menu with disabled item](https://hugecalf-studios.github.io/unity-lessons/lessons/editor/menuitem/disabled.png)


The above is made with the following code. Note how Item1 has a Validation Function assigned which returns false. This will mean Item1 is always disabled. Also note that Item2 has no Validation Function, which acts the same as a Validation Function which returns true, meaning Item2 is always enabled.

```csharp
[MenuItem(Item1MenuPath, priority = 1)]
private static void Item1() {
    Debug.Log("Item 1!");
}

[MenuItem(Item1MenuPath, true)]
private static bool Item1Validate() {
    return false;
}

[MenuItem(Item2MenuPath, priority = 2)]
private static void Item2() {
    Debug.Log("Item 2!");
}
```
